plugins/aws/providers.go

```diff
@@ -0,0 +1,142 @@
+package aws
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/1Password/shell-plugins/sdk"
+	"github.com/1Password/shell-plugins/sdk/schema/fieldname"
+	confighelpers "github.com/99designs/aws-vault/v7/vault"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+)
+
+type assumeRoleProvider struct {
+	confighelpers.AssumeRoleProvider
+	stsCacheWriter
+}
+
+func (p assumeRoleProvider) Retrieve(ctx context.Context) (aws.Credentials, error) {
+	credentials, err := p.AssumeRoleProvider.Retrieve(ctx)
+	if err != nil {
+		return aws.Credentials{}, err
+	}
+
+	err = p.stsCacheWriter.persist(credentials)
+	if err != nil {
+		return aws.Credentials{}, err
+	}
+
+	return credentials, nil
+}
+
+func NewAssumeRoleProvider(awsConfig *confighelpers.Config, in sdk.ProvisionInput, out *sdk.ProvisionOutput) aws.CredentialsProvider {
+	roleCacheKey := getRoleCacheKey(awsConfig.RoleARN)
+	if in.Cache.Has(roleCacheKey) {
+		return NewStsCacheProvider(roleCacheKey, in.Cache)
+	}
+
+	if awsConfig.HasMfaSerial() && awsConfig.MfaToken != "" {
+		return &assumeRoleProvider{
+			AssumeRoleProvider: confighelpers.AssumeRoleProvider{
+				StsClient:         getSTSClient(awsConfig.Region, NewMFASessionTokenProvider(awsConfig, in, out)),
+				RoleARN:           awsConfig.RoleARN,
+				RoleSessionName:   awsConfig.RoleSessionName,
+				ExternalID:        awsConfig.ExternalID,
+				Duration:          900 * time.Second, // minimum duration of 15 minutes
```

Comment:
Can we dictate this? For example, what if the user wants to run `aws logs tail --follow <log-group>`? That command can run for more than 15 minutes.

Could we use the session duration from the config instead to allow the user to configure this?

Reply:
I did not think of such a use case. Indeed it makes sense to allow the user to override this. I still kept the 15 minutes expiration (the minimum allowed one) as a default in case none is specified because this is more secure than the 1 hour default expiration of AWS.

Reply:
Done